File: Csharp/Patterns/LC00107_bfs_medium.cs
using Csharp.LC.Classes;

namespace Csharp.Patterns.LC00107
{
    public class Solution
    {
        public IList<IList<int>> LevelOrderBottom(TreeNode root)
        {
            if (root is null) return new List<IList<int>>();

            Queue<TreeNode> queue = new Queue<TreeNode>();
            List<IList<int>> result = new List<IList<int>>();
            queue.Enqueue(root);

            while (queue.Any())
            {
                int levelSize = queue.Count();
                List<int> level = new List<int>();

                for (int i = 0; i < levelSize; ++i)
                {
                    TreeNode node = queue.Dequeue();
                    level.Add(node.val);

                    if (node.left is not null)
                    {
                        queue.Enqueue(node.left);
                    }
                    if (node.right is not null)
                    {
                        queue.Enqueue(node.right);
                    }
                }
                result.Insert(0, level);
            }
            return result;
        }
    }
}

File: Python/Patterns/LC00107_bfs_medium.py
from collections import deque
from typing import List

def levelOrderBottom(root) -> List[List[int]]:
    if root is None:
        return []

    queue = deque([root])
    result = []

    while queue:
        level_size = len(queue)
        curr_level = []

        for _ in range(level_size):
            node = queue.popleft()
            curr_level.append(node.val)

            if node.left:
                queue.append(node.left)
            if node.right:
                queue.append(node.right)
        result.insert(0, curr_level)
    return result